### PR TITLE
drivers: sensor: lis2dw12: Add Power Domain Handling

### PR DESCRIPTION
**Pull request by SSaeedTonies**
The lis2dw12 accelerometer driver now supports power domains via CONFIG_PM_DEVICE. If the config is selected, the driver will be required to provide a power-domain and will initialize itself by enabling pm_device_runtime on this domain.

Tap detection can be enabled by setting the sensor_trigger_set callback to a valid callback function - which in turn gets the power domain and keeps it until sensor_trigger_set is called with a NULL callback. This way, interrupt based events can be caught, while the driver stays on a power domain - assuming that the detection is disabled once it is no longer needed.

Similarly, a sensor attribute for continuous measurement mode has been added to allow for an application to continuously fetch data from the sensor without having to power - and reinitialize - only during a fetch call. When set using sensor_value, the driver keeps the power domain and does not put it back until the attribute is reset again.

Additional flags have been introduced to handle any overlap of domain requesting/putting back. The ability to set the default fifo mode has also been added.

Clang formatting has also been introduced for previously incorrectly formatted source files

### CODE REVIEW

File: dts/bindings/sensor/st,lis2dw12-common.yaml [hunk 1]
```diff
@@ -104,6 +104,39 @@ properties:
       - 3
       - 4
 
+  fifo-mode:
+    type: int
+    default: 0
+    description: |
+      Specify the sensor FIFO mode. Default is 0
```
Comment (MaureenHelm) [inline dts/bindings/sensor/st,lis2dw12-common.yaml:111]: Please document in the description why this default value was chosen (is it a hardware register reset value?), not repeat what the value is

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for the contribution!

Please separate clang formatting changes from logical changes into different commits

@avisconti please take a look

File: dts/bindings/sensor/st,lis2dw12-common.yaml (hunk 1 above)
Comment (SSaeedTonies) [reply dts/bindings/sensor/st,lis2dw12-common.yaml]: Done :)

File: include/zephyr/drivers/sensor.h [hunk 2]
```diff
@@ -315,6 +315,8 @@ enum sensor_attribute {
 	SENSOR_ATTR_CALIBRATION,
 	/** Enable/disable sensor features */
 	SENSOR_ATTR_FEATURE_MASK,
+	/** Enable/disable continuous mode. */
+	SENSOR_ATTR_CONTINUOUS_MODE,
```
Comment (avisconti) [inline include/zephyr/drivers/sensor.h:319]: I think that adding this feature in sensor.h requires a separate commit

File: dts/bindings/sensor/st,lis2dw12-common.yaml [hunk 3]
```diff
@@ -104,6 +104,39 @@ properties:
...
+        6 # Continuous mode. This mode provides a continuous FIFO update:
+        when 32 unread samples are stored in memory, as new data arrives
+        the oldest data is discarded and overwritten by the newer.
+
+    enum:
+      - 0
+      - 1
+      - 3
+      - 4
+      - 6
```
Comment (avisconti) [inline dts/bindings/sensor/st,lis2dw12-common.yaml:138]: 1. Is this change related to the commit itself? Or maybe should it be in a separate commit?
2. In any case, I don't see how you read the multiple data in FIFO. Shouldn't you change data fetch and channel get? And in which way?

**Review by avisconti — CHANGES_REQUESTED** [review]

File: dts/bindings/sensor/st,lis2dw12-common.yaml (hunk 3 above)
Comment (SSaeedTonies) [reply dts/bindings/sensor/st,lis2dw12-common.yaml:138]: 1. It is not directly related to the commit. I can keep the implementation of fifo-mode selection as a separate commit if wished.
2. Multiple data read is done internally within the driver's FIFO buffer. Here we can merely select which type of "multiple data read" we want. Data fetch and channel get do not need to be changed as the output of the FIFO buffer is still read from the same register address

File: include/zephyr/drivers/sensor.h (hunk 2 above)
Comment (SSaeedTonies) [reply include/zephyr/drivers/sensor.h:319]: I have added this "Continuous Mode" to facilitate the power domain handling in the case of keeping the sensor active for a larger period of time - as such it is directly part of my functionality of introducing power domain to the driver. This change has nothing to do with fifo-mode and my functionality depends on this attribute being defined. Please let me know how to proceed in this case :)

Comment (avisconti) [reply include/zephyr/drivers/sensor.h:319]: the fifo part is a different story.
But regarding the "Continuos Mode" I think that anyway the change to sensor.h must be for generic use, hence in a separate commit and well documented, as they might be used by other drivers as well

File: dts/bindings/sensor/st,lis2dw12-common.yaml (hunk 3 above)
Comment (avisconti) [reply dts/bindings/sensor/st,lis2dw12-common.yaml:138]: > 1. It is not directly related to the commit. I can keep the implementation of fifo-mode selection as a separate commit if wished.
> 

Well, it's not my wish. I think it's mandatory to have atomic commits, with a single feature for each commit.
;)

>     2. Multiple data read is done internally within the driver's FIFO buffer. Here we can merely select which type of "multiple data read" we want. Data fetch and channel get do not need to be changed as the output of the FIFO buffer is still read from the same register address, regardless of the FIFO mode selection.

But let's say you receive an interrupt because you reached FIFO threshold (is this also checked in the driver?), then how can you read ALL data from the FIFO if you don't check number of FIFO items? Do you have a sample (which I guess would be nice to commit) to test it? Maybe you can prepare all this part in a separate PR

Comment (avisconti) [reply dts/bindings/sensor/st,lis2dw12-common.yaml:138]: SPOILER: I'm not sure we can handle FIFO in sensor API v1.

Comment (SSaeedTonies) [reply dts/bindings/sensor/st,lis2dw12-common.yaml:138]: I think I can move the fifo mode select change into a separate PR later. Would be satisfied for now to introduce the power domain handling only :)

**Review by avisconti — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/sensor.h (hunk 2 above)
Comment (SSaeedTonies) [reply include/zephyr/drivers/sensor.h:319]: Change now kept in a separate commit

File: dts/bindings/sensor/st,lis2dw12-common.yaml (hunk 3 above)
Comment (SSaeedTonies) [reply dts/bindings/sensor/st,lis2dw12-common.yaml]: fifo mode select changes now discarded for this PR

Comment (teburd) [reply dts/bindings/sensor/st,lis2dw12-common.yaml]: @avisconti FIFO mode could be supported if the driver moves to the newer API

Comment (avisconti) [reply dts/bindings/sensor/st,lis2dw12-common.yaml]: @SSaeedTonies I think you need to drop the CONTINUOS_MODE idea and propose a PR with only the POWER_MODE support

**Review by avisconti — APPROVED** [review]

**Comment by avisconti** [comment]
Please rebase. Sorry I missed this PR evolution
